### DISCUSSION TITLE
generalized_inner_product() and handling of symmetric tensors

### DISCUSSION THREAD

**Discussion by jiyanshagarwal** [ansys/pydpf-core#1886] [Q&A] [answered]
Hello!

I am working on a script with pyDPF-core and had a question about how generalized_inner_product() works. I am computing the traction vector from the stress vector and surface normal. However, when I use the generalized_inner_product() operator, I got unexpected results.

Made a minimal example:

```
from ansys.dpf import core as dpf

# Initialize Fields
field1 = dpf.Field(nentities=2, nature=dpf.natures.symmatrix)
field2 = dpf.Field(nentities=1)

field1.data = [1, 2, 3, 4, 5, 6, 11, 2, 3, 4, 5, 6]
field2.data = [1, 1, 1]

field1.scoping.ids = range(2)
field2.scoping.ids = range(1)

dot_op = dpf.operators.math.generalized_inner_product(field1, field2)

field3 = dot_op.eval()
print(field3.data)
```

Output:
[[6. 6.5 8.5]
[0. 0. 0. ]]

Expected Output:
[[11. 11. 14.]
[0. 0. 0. ]]

This confirmed that when acting on fields with nature=symmatrix, generalized_inner_product() divides the off-diagonal components by 2 before computing the product.

Wanted to know if this is intended behavior. Is there some property I can set to get around this?

**Comment by mikerife** [comment]
Hi @jiyanshagarwal 
I've looked into this forwards, backwards and sideways a few times and sorry to say I don't have a good answer for you.  I've made an internal request for some developer help...or at least get pointed to some clarifying document.  Just wanted to let you know that someone is at least trying!

Mike

**Reply by jiyanshagarwal** [reply]
Thanks, would love to get some clarity on this.

**Comment by ayush-kumar-423** [comment]
@jiyanshagarwal to get correct matrix multiplication, you can define your stress tensor as a 3x3 matrix and your surface normal as a 3x1 matrix. See example below, let me know if this works for you.

```
from ansys.dpf import core as dpf

f1 = dpf.fields_factory.create_matrix_field(1, 3, 3)
f1.scoping.ids = [1]

# Arrange stress tensor in a 3x3 Matrix
"""
tensor = [S_xx, S_xy, S_xz
          S_yx, S_yy, S_yz
          S_zx, S_zy, S_zz]
"""

f1.data = [  # Tensor 1
    100, 20, 30,
    20, 150, 50,
    30, 50, 200]

f2 = dpf.fields_factory.create_matrix_field(1, 3, 1)
f2.scoping.ids = [1]
f2.data = [1, 2, 3]

dot_op = dpf.operators.math.generalized_inner_product(f1, f2)
dot_op_f = dot_op.outputs.field.get_data()
print(dot_op_f)
```

![image](https://github.com/user-attachments/assets/2898ea5d-d64e-4d86-ae69-452a266a0ec5)

**Reply by mikerife** [reply]
@ayush-kumar-423 
Yes of course we can do that; but the question is why do we need to?  Is there a reference to how PyDPF treats symmatrix enums with respect to the generalized inner product?

**Reply by mikerife** [reply, marked as answer]
@jiyanshagarwal the issue has been fixed but I do not yet have a date for the release where the fix will be available.  In the mean time please implement the method Ayush posted...or another as you see fit.

**Reply by jiyanshagarwal** [reply]
@mikerife Thank you!

**Reply by mikerife** [reply]
@jiyanshagarwal YW.  Looks like the issue was fixed in 25 R2 (not found in time for 25 R1 release).  Mike
